sriovnet.go

```diff
@@ -450,15 +462,55 @@ func GetNetDevicesFromPci(pciAddress string) ([]string, error) {
 	return netDevices, nil
 }
 
-// GetPfPciFromVfPci retrieves the parent PF PCI address of the provided VF PCI address in D:B:D.f format
-func GetPfPciFromVfPci(vfPciAddress string) (string, error) {
-	pfPath := filepath.Join(PciSysDir, vfPciAddress, "physfn")
-	pciDevDir, err := utilfs.Fs.Readlink(pfPath)
+// GetNetDeviceFromAux gets auxiliary device name (e.g 'mlx5_core.sf.2') and
+// returns the correlate netdevice
+func GetNetDevicesFromAux(auxDev string) ([]string, error) {
+	auxDir := filepath.Join(AuxSysDir, auxDev, "net")
+	return getNetDevicesFromPath(auxDir)
+}
+
+// GetNetDevicesFromPci gets a PCI address (e.g '0000:03:00.1') and
+// returns the correlate list of netdevices
+func GetNetDevicesFromPci(pciAddress string) ([]string, error) {
+	pciDir := filepath.Join(PciSysDir, pciAddress, "net")
+	return getNetDevicesFromPath(pciDir)
+}
+
+// GetSfIndexByAuxDev gets a SF device name (e.g 'mlx5_core.sf.2') and
+// returns the correlate SF index.
+func GetSfIndexByAuxDev(auxDev string) (int, error) {
+	sfnumFile := filepath.Join(AuxSysDir, auxDev, "sfnum")
+	if _, err := utilfs.Fs.Stat(sfnumFile); err != nil {
+		return -1, fmt.Errorf("cannot get sfnum for %s device: %v", auxDev, err)
+	}
+
+	sfnumStr, err := utilfs.Fs.ReadFile(sfnumFile)
 	if err != nil {
-		return "", fmt.Errorf("failed to read physfn link, provided address may not be a VF. %v", err)
+		return -1, fmt.Errorf("cannot read sfnum file for %s device: %v", auxDev, err)
 	}
 
-	pf := path.Base(pciDevDir)
+	sfnum, err := strconv.Atoi(string(sfnumStr))
+	if err != nil {
+		return -1, err
+	}
+	return sfnum, nil
+}
+
+// GetPfPciFromAux retrieves the parent PF PCI address of the provided auxiliary device in D.T.f format
+func GetPfPciFromAux(auxDev string) (string, error) {
+	auxPath := filepath.Join(AuxSysDir, auxDev)
+	absoluteAuxPath, err := utilfs.Fs.Readlink(auxPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read auxiliary link, provided device ID may be not auxiliary device. %v", err)
+	}
+
+	parent := filepath.Dir(absoluteAuxPath)
+	pf := filepath.Base(parent)
+	if pf != "" && pf[0:4] != "0000" { // XXX regexp??
+		// it's a nested auxiliary device. repeat
```

Comment:
Aux device can be like
.../0000:03:00.0/mlx5_core.sf.4/mlx5_core.eth.4

So parent device for eth.4 is SF aux device.

Reply:
id like to get more context on, what is sf device type and eth device type, what other devices we might have ? can we have a "deeper" nesting ?

not familiar with this, so appologies if questions are naïve :)

Reply:
As pointed offline by Dmytro: https://www.kernel.org/doc/html/latest/driver-api/auxiliary_bus.html doesn't seem to talk much about nesting.

Also as Dmytro suggested we could loop something like:
```
while isAuxName(fileName) {
  parent = getParent(fileName)
 fileName = filepath.Base(parent)
}

if isPciBus(fileName) {
/* found */
} else {
/* error */
}
```

It will allow to skip any numbers of nested devices up to pci bus or not valid parent device

Reply:
i think its worth exploring the use of devlink port instead of sysfs implementation to avoid this logic.

can there be both eth and rdma under the same sf device ?
i believe its worth understanding this nesting and aux device heirarcies (if you do please explain to me)

feels to me that using netlink/mlxdevm would be simpler and less error prone.

Reply:
> can there be both eth and rdma under the same sf device ?

Yes. IIRC, It'll look like:
`.../0000:03:00.0/mlx5_core.sf.4/mlx5_core.eth.4
.../0000:03:00.0/mlx5_core.sf.4/mlx5_core.ib.4`

> feels to me that using netlink/mlxdevm would be simpler and less error prone.

I've totally forgot to mention that devlink/mlxdevm not aware about auxiliary devices. There is no binding between iface name and aux name available to netlink infrustructure.

Reply:
We will use `devlink/mlxdevm` in an incremental commit